### PR TITLE
WIP: Upgrade to Solarium 6

### PR DESCRIPTION
**Pull request by japerry**
Testing the Solarium 6 upgrade. Making PR now to test actions.

### CODE REVIEW

File: search_api_solr.install [hunk 1]
```diff
@@ -19,24 +19,22 @@ function search_api_solr_requirements($phase) {
     }
   }
   elseif ($phase == 'runtime') {
-    if (strtotime('1 January 2021') < time()) {
+    if (\Drupal::moduleHandler()->moduleExists('acquia_search')) {
+      $ret['search_api_solr_eol'] = [
+        'title' => t('Search API Solr 8.x-1.x'),
+        'value' => t('1.x LTS Support'),
+        'description' => t('Community support of Search API Solr 1.x Ends 2020-12-31. Please contact Acquia support to report any issues with this module.'),
```
Comment (mkalkbrenner) [inline search_api_solr.install]: This is not true. The community support has been extended throughout 2021.

File: src/Solarium/Autocomplete/Query.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+use Solarium\Component\Suggester;
+use Solarium\Component\Terms;
+use Solarium\Core\Query\AbstractQuery;
+use Solarium\Core\Query\RequestBuilderInterface;
+use Solarium\Core\Query\ResponseParserInterface;
+
+/**
+ * Autocomplete query.
+ */
+class Query extends AbstractQuery implements ComponentAwareQueryInterface {
```
Comment (mkalkbrenner) [inline src/Solarium/Autocomplete/Query.php]: This class is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/Solarium/Autocomplete/RequestBuilder.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+namespace Drupal\search_api_solr\Solarium\Autocomplete;
+
+use Solarium\Core\Client\Request;
+use Solarium\Core\Query\AbstractQuery;
+use Solarium\Core\Query\AbstractRequestBuilder;
+
+/**
+ * Autocomplete request builder.
+ */
+class RequestBuilder extends AbstractRequestBuilder {
```
Comment (mkalkbrenner) [inline src/Solarium/Autocomplete/RequestBuilder.php]: This class is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/Solarium/Autocomplete/ResponseParser.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+namespace Drupal\search_api_solr\Solarium\Autocomplete;
+
+use Solarium\Core\Query\AbstractResponseParser;
+use Solarium\Core\Query\ResponseParserInterface;
+use Solarium\Core\Query\Result\ResultInterface;
+
+/**
+ * Autocomplete response parser.
+ */
+class ResponseParser extends AbstractResponseParser implements ResponseParserInterface {
```
Comment (mkalkbrenner) [inline src/Solarium/Autocomplete/ResponseParser.php]: This class is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/Solarium/Autocomplete/Result.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace Drupal\search_api_solr\Solarium\Autocomplete;
+
+use Solarium\Component\ComponentAwareQueryInterface;
+use Solarium\Core\Query\Result\QueryType as BaseResult;
+
+/**
+ * Autocomplete query result.
+ */
+class Result extends BaseResult {
```
Comment (mkalkbrenner) [inline src/Solarium/Autocomplete/Result.php]: This class is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/Solarium/Result/StreamDocument.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace Drupal\search_api_solr\Solarium\Result;
+
+use Solarium\Core\Query\AbstractDocument;
+use Solarium\Core\Query\DocumentInterface;
+
+/**
+ * Stream result Solr document.
+ */
+class StreamDocument extends AbstractDocument {
```
Comment (mkalkbrenner) [inline src/Solarium/Result/StreamDocument.php]: This class is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 7]
```diff
@@ -15,6 +17,25 @@
  */
 interface SolrBackendInterface extends BackendInterface {
 
+  /**
+   * The minimum required Solr schema version.
+   */
+  const SEARCH_API_SOLR_MIN_SCHEMA_VERSION = '4.1.1';
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: Not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 8]
```diff
@@ -15,6 +17,25 @@
...
+   * We must not use any character that has a special meaning within regular
+   * expressions. Additionally we have to avoid characters that are valid for
+   * Drupal machine names.
+   * The end of a language ID is indicated by an underscore '_' which could not
+   * occur within the language ID itself because Drupal uses lanague tags.
+   *
+   * @see http://de2.php.net/manual/en/regexp.reference.meta.php
+   * @see https://www.w3.org/International/articles/language-tags/
+   */
+  const SEARCH_API_SOLR_LANGUAGE_SEPARATOR = ';';
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 9]
```diff
@@ -27,14 +48,51 @@ interface SolrBackendInterface extends BackendInterface {
...
+   *   The Search API index entity.
+   * @param bool $reset
+   *   (optional) Whether to reset the static cache.
+   *
+   * @return array
+   *   The language-specific mapping from Drupal to Solr field names.
+   *
+   * @throws \Drupal\search_api\SearchApiException
+   */
+  public function getLanguageSpecificSolrFieldNames($language_id, ?IndexInterface $index, $reset = FALSE);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 10]
```diff
@@ -27,14 +48,51 @@ interface SolrBackendInterface extends BackendInterface {
...
+   *   The Search API index entity.
+   * @param bool $reset
+   *   (optional) Whether to reset the static cache.
+   *
+   * @return array
+   *   The language-specific mapping from Drupal to Solr field names.
+   *
+   * @throws \Drupal\search_api\SearchApiException
+   */
+  public function getSolrFieldNamesKeyedByLanguage(array $language_ids, IndexInterface $index, $reset = FALSE);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 11]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
    */
   public function extractContentFromFile($filepath);
 
+  /**
+   * Returns the targeted content domain of the server.
+   *
+   * @return string
+   *   The content domain.
+   */
+  public function getDomain();
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 12]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+
+  /**
+   * Returns the targeted environment of the server.
+   *
+   * @return string
+   *   The environment.
+   */
+  public function getEnvironment();
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 13]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+
+  /**
+   * Indicates if the Solr server uses a managed schema.
+   *
+   * @return bool
+   *   TRUE if the Solr server uses a managed schema, FALSE if the Solr server
+   *   uses a classic schema.
+   */
+  public function isManagedSchema();
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 14]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+
+  /**
+   * Indicates if the Solr index should be optimized daily.
+   *
+   * @return bool
+   *   TRUE if the Solr index should be optimized daily, FALSE otherwise.
+   */
+  public function isOptimizeEnabled();
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 15]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+  /**
+   * Returns a ready to use query string to filter results by index and site.
+   *
+   * @param \Drupal\search_api\IndexInterface $index
+   *   The Search API index entity.
+   *
+   * @return string
+   *   The query string filter.
+   */
+  public function getIndexFilterQueryString(IndexInterface $index);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 16]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   * In case of Solr Cloud an index might use a different Solr collection.
+   *
+   * @param \Drupal\search_api\IndexInterface $index
+   *
+   * @return \Solarium\Core\Client\Endpoint
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   * @throws \Drupal\search_api\SearchApiException
+   */
+  public function getCollectionEndpoint(IndexInterface $index);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 17]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   *
+   * @param \Drupal\search_api\IndexInterface $index
+   *   The Search API index entity.
+   *
+   * @return array
+   *   An associative array of settings.
+   *
+   * @deprecated use \Drupal\search_api_solr\Utility\Utility::getIndexSolrSettings()
+   */
+  public function getIndexSolrSettings(IndexInterface $index);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 18]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   * - If set, the index-specific prefix.
+   * - The index's machine name.
+   *
+   * @param \Drupal\search_api\IndexInterface $index
+   *   The Search API index entity.
+   *
+   * @return string
+   *   The prefixed machine name.
+   */
+  public function getIndexId(IndexInterface $index);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 19]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   *
+   * @param \Drupal\search_api\IndexInterface $index
+   *   The Search API index entity.
+   *
+   * @return string
+   *   The targeted Index ID.
+   *
+   * @throws \Drupal\search_api\SearchApiException
+   */
+  public function getTargetedIndexId(IndexInterface $index);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 20]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   *
+   * @param \Drupal\search_api\IndexInterface $index
+   *   The Search API index entity.
+   *
+   * @return string
+   *   The targeted site hash.
+   *
+   * @throws \Drupal\search_api\SearchApiException
+   */
+  public function getTargetedSiteHash(IndexInterface $index);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 21]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   * @param \Drupal\search_api\Query\QueryInterface $query
+   *   The query used for the streaming expression.
+   *
+   * @return \Solarium\QueryType\Stream\Result
+   *   The streaming expression result.
+   *
+   * @throws \Drupal\search_api\SearchApiException
+   * @throws \Drupal\search_api_solr\SearchApiSolrException
+   */
+  public function executeStreamingExpression(QueryInterface $query);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 22]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   * @param \Drupal\search_api\Query\QueryInterface $query
+   *   The query used for the graph streaming expression.
+   *
+   * @return \Solarium\QueryType\Graph\Result
+   *   The graph streaming expression result.
+   *
+   * @throws \Drupal\search_api\SearchApiException
+   * @throws \Drupal\search_api_solr\SearchApiSolrException
+   */
+  public function executeGraphStreamingExpression(QueryInterface $query);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 23]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   *
+   * @return bool
+   *   TRUE if a finalization run, FALSE otherwise. FALSE doesn't indicate an
+   *   error!
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   * @throws \Drupal\search_api\SearchApiException
+   * @throws \Drupal\search_api_solr\SearchApiSolrException
+   */
+  public function finalizeIndex(IndexInterface $index);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 24]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   * @return array
+   *   Stats as associative array keyed by language IDs. The value is the
+   *   language id of the corresponding field type existing on the server's
+   *   current schema or FALSE.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   * @throws \Drupal\search_api\SearchApiException
+   * @throws \Drupal\search_api_solr\SearchApiSolrException
+   */
+  public function getSchemaLanguageStatistics(?Endpoint $endpoint = NULL);
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 25]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   * Get document counts for this server, in total and per site / index.
+   *
+   * @return array
+   *   An associative array of document counts.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   * @throws \Drupal\search_api\SearchApiException
+   * @throws \Drupal\search_api_solr\SearchApiSolrException
+   */
+  public function getDocumentCounts();
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 26]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+   * _version_ numbers are important for replication and checkpoints.
+   *
+   * @return array
+   *   An associative array of max document versions.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\PluginException
+   * @throws \Drupal\search_api\SearchApiException
+   * @throws \Drupal\search_api_solr\SearchApiSolrException
+   */
+  public function getMaxDocumentVersions();
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 27]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+
+  /**
+   * Gets a list of Solr Field Types that are disabled for this backend.
+   *
+   * @return String[]
+   */
+  public function getDisabledFieldTypes(): array;
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 28]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+
+  /**
+   * Gets a list of Solr Caches that are disabled for this backend.
+   *
+   * @return String[]
+   */
+  public function getDisabledCaches(): array;
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 29]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+
+  /**
+   * Gets a list of Solr Request Handlers that are disabled for this backend.
+   *
+   * @return String[]
+   */
+  public function getDisabledRequestHandlers(): array;
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 30]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+
+  /**
+   * Gets a list of Solr Request Dispatchers that are disabled for this backend.
+   *
+   * @return String[]
+   */
+  public function getDisabledRequestDispatchers(): array;
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrBackendInterface.php [hunk 31]
```diff
@@ -76,7 +138,248 @@ public function getDocuments(IndexInterface $index, array $items, UpdateQuery $u
...
+
+  /**
+   * Indicates if the the current Solr config should not be verified.
+   *
+   * @return bool
+   */
+  public function isNonDrupalOrOutdatedConfigSetAllowed(): bool;
```
Comment (mkalkbrenner) [inline src/SolrBackendInterface.php]: This is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

File: src/SolrCloudConnectorInterface.php [hunk 32]
```diff
@@ -0,0 +1,134 @@
...
+namespace Drupal\search_api_solr;
+
+use Solarium\Core\Client\Endpoint;
+use Solarium\QueryType\Graph\Query as GraphQuery;
+use Solarium\QueryType\Stream\Query as StreamQuery;
+
+/**
+ * The Solr Cloud connector interface.
+ */
+interface SolrCloudConnectorInterface extends SolrConnectorInterface {
```
Comment (mkalkbrenner) [inline src/SolrCloudConnectorInterface.php]: This class is not needed to migrate Search API Solr 8.x-1.x to Solarium 6.

**Review by mkalkbrenner — CHANGES_REQUESTED** [review]
Most things in this draft are not required for porting Search API Solr 8.x-1.x to Solarium 6.x.

**Comment by narendradesai** [comment]
Thank you @mkalkbrenner , your comments were very helpful. Will update PR accordingly.

**Pull request closed without merging**